common/wireaddr.c

```diff
@@ -10,6 +10,174 @@
 #include <sys/types.h>
 #include <wire/wire.h>
 
+
+#include <common/socksd.h>
+#include <common/base64.h>
+
+
+char *
+b64_encode (const unsigned char *src, size_t len) {
+  int i = 0;
+  int j = 0;
+  char *enc = NULL;
+  size_t size = 0;
+  unsigned char buf[4];
+  unsigned char tmp[3];
+
+  // alloc
+  enc = (char *) b64_malloc(1);
+  if (NULL == enc) { return NULL; }
+
+  // parse until end of source
+  while (len--) {
+    // read up to 3 bytes at a time into `tmp'
+    tmp[i++] = *(src++);
+
+    // if 3 bytes read then encode into `buf'
+    if (3 == i) {
+      buf[0] = (tmp[0] & 0xfc) >> 2;
+      buf[1] = ((tmp[0] & 0x03) << 4) + ((tmp[1] & 0xf0) >> 4);
+      buf[2] = ((tmp[1] & 0x0f) << 2) + ((tmp[2] & 0xc0) >> 6);
+      buf[3] = tmp[2] & 0x3f;
+
+      enc = (char *) b64_realloc(enc, size + 4);
+      for (i = 0; i < 4; ++i) {
+        enc[size++] = b64_table[buf[i]];
+      }
+
+      i = 0;
+    }
+  }
+
+  // remainder
+  if (i > 0) {
+    // fill `tmp' with `\0' at most 3 times
+    for (j = i; j < 3; ++j) {
+      tmp[j] = '\0';
+    }
+
+    // perform same codec as above
+    buf[0] = (tmp[0] & 0xfc) >> 2;
+    buf[1] = ((tmp[0] & 0x03) << 4) + ((tmp[1] & 0xf0) >> 4);
+    buf[2] = ((tmp[1] & 0x0f) << 2) + ((tmp[2] & 0xc0) >> 6);
+    buf[3] = tmp[2] & 0x3f;
+
+    // perform same write to `enc` with new allocation
+    for (j = 0; (j < i + 1); ++j) {
+      enc = (char *) b64_realloc(enc, size + 1);
+      enc[size++] = b64_table[buf[j]];
+    }
+
+    // while there is still a remainder
+    // append `=' to `enc'
+    while ((i++ < 3)) {
+      enc = (char *) b64_realloc(enc, size + 1);
+      enc[size++] = '=';
+    }
+  }
+
+  // Make sure we have enough space to add '\0' character at end.
+  enc = (char *) b64_realloc(enc, size + 1);
+  enc[size] = '\0';
+
+  return enc;
+}
+
+
+#define BASE32_DATA "abcdefghijklmnopqrstuvwxyz234567"
+
+char *b32_encode(char *, u8 *) ;
+
+
+char *b32_encode(char *dst, u8 *src) {
+  u16 byte = 0,  
+          poff = 0; 
+  for(; byte < 	16; poff += 5) {
+    if(poff > 7) {
+      poff -= 8;
+      src++;
+    }
+    dst[byte++] = BASE32_DATA[ (htobe16(*(u16*)src) >> (11 -poff)) & (u16)0x001F];
+  }
+  dst[byte] = 0;
+  return dst;
+}
+
+
+u8 bit_pos(int ,u8 *);
+
+u8 bit_pos(int n,u8 *src)
+{
+int d,i;
+d = n/8;
+i=n - 8+d;
+
+if (src[d] & (1>>i)) return 1;
+return 0;  	
+};
+
+char *b32_encode2(char *dst,u8 *src);
+
+char *b32_encode2(char *dst,u8 *src)
+{
+	int i,j,n;
+	u8 enc;
+	n=0;
+	
+	for(i=0;i< ((10*8)/5);i++)
+ 	{
+    enc=0;
+ 	for (j=0;j<5;j++)
+ 	{
+    enc = enc*2 + bit_pos(n,src);
+    n=n+1;
+ 	}
+  	dst[i]=BASE32_DATA[enc];	
+	};
+	
+	dst[i]=0;
+return dst ;	
+};
+
+//FIXME quiknditry
+int b32_decode( u8 *dst,u8 *src);
+
+int b32_decode( u8 *dst,u8 *src) {
+ 
+  int rem = 0;
+ 
+  int i;
+  u8 *p=src;
+  int buf;
+  u8 ch;
+  
+  for (i=0; i < 16 ; p++) {
```

Comment:
Truly fixed size?  `p` is incremented here but the loop variable is `i`.  You do increment `i` later in the code but maybe better to put the `p` increment in the loop body and the `i` increment here.

Reply:
this is tricky the magic is the increment of (i)  on dep of bits > 111..... 
so (i) i cant be in the main loop but the next char has to be fetched.